e2e/actions.go

```diff
@@ -0,0 +1,120 @@
+package e2e
+
+import (
+	"fmt"
+	"log"
+	"math/rand"
+	"time"
+)
+
+type Scenario struct {
+	actions []Action
+}
+
+func NewScenario() *Scenario {
+	return &Scenario{
+		actions: make([]Action, 0),
+	}
+}
+
+func (s *Scenario) AddAction(action Action) *Scenario {
+	s.actions = append(s.actions, action)
+	return s
+}
+
+func (s *Scenario) CleanUp(cluster *Cluster) {
+	for _, action := range s.actions {
+		action.Revert(cluster)
+	}
+}
+
+type Action interface {
+	Apply(c *Cluster)
+	Revert(c *Cluster)
+}
+
+// This action encapsulates logic for dropping nodes action.
+type DropNodeAction struct {
+	dropCount                int
+	dropProbabilityThreshold int
+	// TODO: Idea: Extract these two fields to another action (called PeriodicAction), which would receive as a parameter another action and would run it?
+	droppingInterval time.Duration
+	duration         time.Duration
+}
+
+func NewDropNodeAction(dropCount, dropProbabilityThreshold int, droppingInterval, duration time.Duration) *DropNodeAction {
+	return &DropNodeAction{
+		dropCount:                dropCount,
+		dropProbabilityThreshold: dropProbabilityThreshold,
+		droppingInterval:         droppingInterval,
+		duration:                 duration,
+	}
+}
+
+func (dn *DropNodeAction) Validate(c *Cluster) error {
+	nodesCount := len(c.nodes)
+	if dn.dropCount >= nodesCount {
+		return fmt.Errorf("trying to drop more nodes (%d) than available in the cluster (%d)", dn.dropCount, nodesCount)
+	}
+	runningNodes := c.GetRunningNodes()
+	nodesLeft := len(runningNodes) - dn.dropCount
+	maxFaultyNodes := GetMaxFaultyNodes(nodesCount)
+	if nodesLeft < maxFaultyNodes {
+		return fmt.Errorf("dropping %d nodes would jeopardize Byzantine fault-tollerant conditions.\nExpected at least %d nodes to run, but action would leave it to %d",
+			dn.dropCount, maxFaultyNodes, nodesLeft)
+	}
+	return nil
+
+}
+
+func (dn *DropNodeAction) Apply(c *Cluster) {
+	if err := dn.Validate(c); err != nil {
+		log.Printf("[WARNING] Skipping drop node action. Reason: '%s'", err)
+		return
+	}
+
+	maxFaultyNodes := GetMaxFaultyNodes(len(c.nodes))
+	ticker := time.NewTicker(dn.droppingInterval)
+	after := time.After(dn.duration)
```

Comment:
I would prefer to have the actions as more functional objects. For example, `Apply()` only checks if its possible to drop a node and drops it. It might return a callback to revert this specific drop. Then, it is up to a higher level function like the `runner` to apply this loop that applies and reverts actions.

Otherwise, as it is right now, you are going to start a lot of concurrent operations that might produce the type of lock issues you have been experiencing. Specially, when the behaviour of executing an action with a certain probability can be easily extracted to the runner. The actions are simplified even more.

Reply:
As I have stated in the https://github.com/0xPolygon/pbft-consensus/pull/24#discussion_r805644418, I agree that periodical logic execution should be removed from `DropNodeAction` and let some higher-level abstraction take care of that part instead.
Regarding probabilistic execution of actions, just want to make sure that I have understood correctly: `Runner` should execute randomly picked `Action` with certain probability (`DropNodeAction` should not take care of that part either). Is that correct?

However is it ok to keep `Action` struct, since it seems as a cleaner solution to me, than having a bunch of functions with a callback revert function as a parameter? Maybe you can take a look at the PR https://github.com/0xPolygon/pbft-consensus/pull/25, since it depends on this one, if we are going to change the way we are representing actions.